Add support for simultaneous runs of Script helper

pnbruckner opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The Script helper (i.e., `homeassistant.helpers.script`) is used by several integrations to implement Home Assistant's [Script Syntax](https://www.home-assistant.io/docs/scripts/), the most notable being `automation` & `script`. However the current behavior, when run multiple times simultaneously (e.g., via an automation that is triggered quickly), is at best unexpected, and worse, requires clumsy workarounds. (More on that below.)

This PR is the first step towards resolving those issues. It improves the Script helper to properly support multiple, simultaneous runs, with options to provide usage flexibility. However, by default, it implements the current behavior to avoid breakage. The intention is to ultimately remove the so-called "legacy" behavior.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #31364 
- Link to documentation pull request: 

### Background
See PR #31740. It was a first attempt at resolving these issues and contains a description of the problems related to the current behavior of the Script helper, as well as YAML for an example automation that is affected by it. However it was decided to break up the changes into smaller chunks, this being the first.

Although not mentioned in the other PR, the typical workaround is to move the automation's `action` sequence into a script, and then in the automation's `action` section invoke the new script like this:
```yaml
action:
- service: script.turn_off
  entity_id: script.new
- service: script.new
```
This is an attempt to implement the new `restart` behavior described below.

Also, there currently isn't an easy way to call a script completly "synchronously" if it contains a `delay` or `wait_template`. The typical workaround to effectively make a script synchronous is to use an `input_boolean`. It would first be turned on by the "caller", then the caller would use a `wait_template` to wait for it to be turned off, which would done by the last step of the called script. This is an attempt to implement the new `blocking` behavior described below.

### New Script helper options
#### if_running
- Controls what happens if Script run (i.e., `async_run` called) while previous run has not yet completed.
- Values:
  - `error` Raise an exception
  - `ignore` Return without doing anything (previous run continues as-is)
  - `parallel` Start run in new task
  - `restart` Stop previous run before starting new run
#### run_mode
- Controls when call to `async_run` will return.
- Values:
  - `background` Returns immediately
  - `legacy` Implements previous behavior, which is to return when done, or when suspended by `delay` or `wait_template`, whichever comes first
  - `blocking` Returns when run has completed
#### Defaults
- If neither is specified, default is `run_mode=legacy` (and `if_running` is not used.)
- Otherwise, defaults are `if_running=parallel` and `run_mode=blocking`.
> Note: If `run_mode` is set to `legacy` then `if_running` must be `None`.


## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

pnbruckner committed:
Add tests for legacy Script helper behavior

pnbruckner committed:
Add Script helper if_running and run_mode options

- if_running controls what happens if Script run while previous run
  has not completed. Can be:
  - error: Raise an exception
  - ignore: Return without doing anything (previous run continues as-is)
  - parallel: Start run in new task
  - restart: Stop previous run before starting new run
- run_mode controls when call to async_run will return. Can be:
  - background: Returns immediately
  - legacy: Implements previous behavior, which is to return when done,
            or when suspended by delay or wait_template
  - blocking: Returns when run has completed
- If neither is specified, default is run_mode=legacy (and if_running
  is not used.) Otherwise, defaults are if_running=parallel and
  run_mode=background. If run_mode is set to legacy then if_running must
  be None.
- Caller may supply a logger which will be used throughout instead of
  default module logger.
- Move Script running state into new helper classes, comprised of an
  abstract base class and two concrete clases, one for legacy behavior
  and one for new behavior.
- Remove some non-async methods, as well as call_from_config which has
  only been used in tests.
- Adjust tests accordingly.

pnbruckner commented:
Tests were added to help make sure the "legacy" behavior is maintained. Test for the new features (i.e., if_running & run_mode) still need to be added.

balloob commented on homeassistant/components/automation/__init__.py:
```diff
@@ -395,10 +395,8 @@ def referenced_entities(self):
-                _LOGGER, f"Error while executing automation {self.entity_id}", err
-            )
+        except Exception:  # pylint: disable=broad-except
+            pass
```
We shouldn't swallow all exceptions. We should catch the ones we expect and handle correctly (like ignoring with `pass`). We should not go and catch everything, as this will swallow attribute errors, type errors etc that are real bugs.

pnbruckner replied:
This is effectively the way it was, except that the logging is now done entirely within the script helper, as opposed to requiring the "user" (in this case, the automation integration) having to call back into the script helper to get the error logged.

To put that another way, the automation integration was already catching all exceptions and just logging them. I did not change that behavior.

Now, having said that, if you still want to change the behavior and have the automation integration not catch anything, it doesn't really matter to me either way, and I can certainly do that.

balloob replied:
The reason we want logging to be called from Automation/Script is because that means we give the caller control about logging/error handling, and the caller can also add a reference to who owns the script (Automation X got an exception, Script Y got an exception)

pnbruckner replied:
The idea you had suggested before was to create a logger specific to the automation or script, e.g., `logging.getLogger(f"{__name__}.{self.name}")` and pass that into the script helper to use. That would satisfy that need. (E.g., the logger might be `homeassistant.components.automation.hello` instead of just `homeassistant.components.automation`.) I was trying to keep the changes in this PR limited to the script helper, while only changing as little elsewhere as possible. That's why I haven't yet fully implemented your earlier suggestion.

In the end, the script helper takes care of logging the exception (on the caller's behalf, by using the logger it passed in) rather than requiring the caller to make a callback into the script helper to log the error appropriately. That is much cleaner and much less error prone (because now there is no longer the issue of the saved exception state, like what step it was on, etc., getting lost.) And since the exception is re-raised, the caller still has the option to decide what to do with it (other than logging, which has been done for it.)

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -59,6 +56,30 @@
+# First choice is default
+RUN_MODE_CHOICES = [
+    RUN_MODE_BACKGROUND,
+    RUN_MODE_BLOCKING,
```
Blocking should be the default.

pnbruckner replied:
Remember, this is not just used by automations. It is also used by the script integration as well as a few others. The automation integration can define a different default for itself (via its config schema) and pass that in, effectively overriding the script helper's default. I chose background as the default specifically to try and be as compatible with the behavior of other integrations (besides automation & script) as possible. But, again, doesn't really matter to me.

balloob replied:
I think that we should pick the default that we want future integrations to use. I would expect that executing a script will take as long as the script takes to execute.

Like you said, we can change the default for script/automation to match legacy behavior for now.

pnbruckner replied:
Ok, fair enough. I'll change the default to blocking.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -121,180 +132,82 @@ class _SuspendScript(Exception):
             try:
-                await self._handle_action(action, variables, context)
+                await getattr(self, f"_async_{_determine_action(self._action)}_step")()
             except _SuspendScript:
```
No need to catch it if you're just raising it.

pnbruckner replied:
Actually, there is, because otherwise it gets caught by the `except Exception` clause that follows.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -121,180 +132,82 @@ class _SuspendScript(Exception):
-        if self._change_listener:
-            self.hass.async_add_job(self._change_listener)
+    def _stop(self):
+        self._script._runs.remove(self)  # pylint: disable=protected-access
```
Why have this? Shouldn't we just have the script register itself as a change_listener?

pnbruckner replied:
This code evolved quite a bit as I struggled to reorganize / clean things up, while at the same time still being able to support the "legacy" behavior. I found this the cleanest way. Besides, the change_listener is one function/method passed into Script. I didn't extend the concept by allowing multiple change listeners, or to somehow use that mechanism between Script and _ScriptRunXXX. Bottom line, when a script run completes, or is told to stop, it needs to be removed from the list of "runs", and I found this the cleanest way to do it.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -121,180 +132,82 @@ class _SuspendScript(Exception):
+                self._log_exception(err)
+                self._stop()
                 # Pass exception on.
                 raise
```
Change listeners are not getting called now.

pnbruckner replied:
That's the way it was. I was trying not to change "legacy" behavior. I suppose that since the exception it getting raised to the calling integration, it can do that itself it if wants (i.e., whatever the "change listener" function/method would have done that it passed into the script helper.) Is that ok, or do you want me to call the change listener before re-raising the exception?

balloob replied:
True. I think it's a bug. It seems like the state can get out of sync now ?

pnbruckner replied:
I definitely looked at the current behavior as a bug, or at least a deficiency. I think it's cleaner, too, if the change listener is called no matter why the run has stopped. I'll change it.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -309,173 +222,288 @@ def async_log_exception(self, logger, message_base, exception):
+        self._script._raise(msg, *args, exception=exception)
+
+
+class _ScriptRun(_ScriptRunBase):
```
Let's name each class to the type it implements. 
```suggestion
class _AsynchronousScriptRun(_ScriptRunBase):
```

pnbruckner replied:
There's only "legacy" (_LegacyScriptRun) and non-legacy (_ScriptRun). Background & blocking run modes both use _ScriptRun, because it's a function of the caller, not the callee.

And, ultimately, I think that the legacy stuff should get ripped out (after a reasonable deprecation warning period.) At that time _ScriptRunBase can be merged into _ScriptRun to simplify the implementation. I only factored out the common stuff into a base class to make adding legacy support easier and cleaner.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -309,173 +222,288 @@ def async_log_exception(self, logger, message_base, exception):
+class _ScriptRun(_ScriptRunBase):
+    """Manage asynchronous run of Script sequence."""
+
+    def __init__(
```
You can also drop the `__init__` and just add this
```suggestion
    task: Optional[asyncio.Task] = None
```

pnbruckner replied:
I suppose. I've never used that style of coding. But, then again, I'm still relatively new to Python. (I coded in C++ much more extensively.) Does it have any impact on performance, or is it just a style issue?

pnbruckner replied:
Ah, ok. I read up on class attributes vs instance attributes and this makes sense. Will do.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -485,26 +513,203 @@ def async_script_timeout(now):
+            self._runs.append(run)
+            await run.async_run()
+        else:
+            run = cast(_ScriptRunBase, _ScriptRun(self._hass, self, variables, context))
```
Is this `cast` necessary? Shouldn't inheritance know this?

pnbruckner replied:
You'd think, but apparently pylint, mpyp etc. don't understand that. I struggled to minimize the amount of crazy casting and typing I had to add to get the check tools to stop complaining.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -485,26 +513,203 @@ def async_script_timeout(now):
+        else:
+            run = cast(_ScriptRunBase, _ScriptRun(self._hass, self, variables, context))
+            self._runs.append(run)
+            cast(_ScriptRun, run).task = self._hass.async_create_task(run.async_run())
```
This seems very weird. I would expect `run.async_run()` would just set its task with `self.task = asyncio.current_task()`

pnbruckner replied:
Ah! I looked and Googled and ... trying to find `asyncio.current_task()`. THANK YOU!! (RTOS's I've used had a call like that, but I couldn't find an equivalent in `asyncio`.) That should certainly simplify things. Let me work on that, and if I can simplify, I'll add a new commit.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -309,173 +222,288 @@ def async_log_exception(self, logger, message_base, exception):
+
+    def stop(self) -> None:
+        """Stop script run."""
+        cast(asyncio.Task, self.task).cancel()
```
This can be `None`, seems dangerous to just cast it.

pnbruckner replied:
I'm also relatively new to typing in Python. I can tell you I had to make quite a few changes to make pylint, mypy, etc. happy. I actually think it makes the code _less_ clear. But that's neither here nor there. The main point is, no, it can't be None. For the non-legacy _ScriptRun, there will _always_ be a Task in `self.task` that `async_run`, etc. run in. That's part of the design, so that things can get cleaned up if told to stop.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -485,26 +513,203 @@ def async_script_timeout(now):
+            run = cast(_ScriptRunBase, _ScriptRun(self._hass, self, variables, context))
+            self._runs.append(run)
+            cast(_ScriptRun, run).task = self._hass.async_create_task(run.async_run())
+            if self._run_mode == RUN_MODE_BLOCKING:
```
I don't understand this, I would expect this logic to be in the `_ScriptRun` inherited classes. For `_BlockingScriptRun` it would just run it, while in `_AsynchronousScriptRun` the `async_run` method would schedule it and store the task.

pnbruckner replied:
Well, the way I looked at it, _what_ `_ScriptRun.async_run` did was independent of whether or not the caller waited for it to finish. That's why I implemented it the way I did. But due to not being able to find `asyncio.current_task`, the implementation was probably more complicated than it needed to be (and used one extra Task in the blocking case than it needed to.) Again, let me work on it...

balloob replied:
When I first read your code, I was totally expecting different `_ScriptRun` inherited classes so you can handle the running differently in `async_run`. I understand now that the base class was only to have a new one just for legacy behavior. I think that we could still be cleaner if we have different classes. Also would make testing easier as it's easy to test those different classes independently.

pnbruckner replied:
Yeah, good point. I hadn't thought of it. I'll add it to the list for the re-work & next commit. Thanks!

balloob commented:
This is off to a really great start! 🎉

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -121,180 +132,82 @@ class _SuspendScript(Exception):
-
-        for cur, action in islice(enumerate(self.sequence), self._cur, None):
+    async def _async_run(self, start):
+        for self._step, self._action in islice(
```
We should add a `self._should_stop` that is an asyncio.Event.

At the start of each loop we should check if the variable exists, stop and set the event if it is. That way the `_BlockingScriptRun` can have their `stop` method stop gracefully (after the current action) instead of cancelling the running running task. What do you think of that?

Now stop can be implemented as 

```
async def async_stop(self):
    self._should_stop = asyncio.Event()
    await self._should_stop.wait()
```

pnbruckner replied:
Interesting idea. I did struggle a bit with the idea that a run could get stopped no matter what it was currently doing. And since it could be calling into who-knows-what integration...

It does make a lot of sense to have a more controlled stop.

Which makes me think again about timeouts. As I think I mentioned in a comment in the previous PR, what to do about the service call's 10 second limit???

If I understand correctly how it works, the call will return after no more than 10 seconds, however the callee isn't canceled. So now how do we handle scripts that can take longer than 10 seconds? I suppose there may have been a way for that to be true before, but given the behavior of suspending for delays & waits, it was not very likely. But now it will be very likely. I guess that's something for one of the follow up PRs to address.

In any case, I'll try to work this suggestion into the re-work using `asyncio.current_task`.

balloob replied:
I think that it's for a follow-up PR, but I think that we could make the timeout overridable/disable the timeout.

pnbruckner committed:
Change per review

- Change run_mode default from background to blocking.
- Make sure change listener is called, even when there's an unexpected
  exception.
- Make _ScriptRun.async_stop more graceful by using an asyncio.Event for
  signaling instead of simply cancelling Task.
- Subclass _ScriptRun for background & blocking behavior.

Also:

- Fix timeouts in _ScriptRun by converting timedeltas to float seconds.
- General cleanup.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -121,6 +133,428 @@ class _SuspendScript(Exception):
+        try:
+            await getattr(self, f"_async_{_determine_action(self._action)}_step")()
+        except Exception as err:
+            if not isinstance(err, (_SuspendScript, _StopScript)):
```
We shouldn't log the exception when it's a BlockingScriptRun either, as the exception will propagate to the caller.

balloob replied:
Maybe a new method, `should_propagate_exception` should be added. A legacy script run would want to not log if it is seeing an exception prior to a delay, because it will propagate up.

pnbruckner replied:
I actually think there are two different issues here. The first is who should log the exception, and the second is whether or not the exception should be propagated.

As to the first, I still think it's cleaner to have the script helper log the error rather than requiring the "user" to call back into it (for the same reasons I gave earlier.)

However, I think you have a point on the second issue. It seems clear that if the "user" is still waiting on the script helper then the exception should be propagated up. It also seems clear that if the "user" is _not_ still waiting that the exception should _not_ be propagated.

I'll work on addressing the second point. Let me know if you still agree on the first.

pnbruckner replied:
It turned out to be pretty easy & clean to control when exceptions should be propagated. 😃

balloob replied:
I think we should think in "who owns the exception handling". If we're in a blocking call, we should propagate. If we are in a standalone task, the script run should handle it. It is totally fine for automation (or any other future caller) to call back to script to log it, because at least we then give automations the option not to.

pnbruckner replied:
>I think we should think in "who owns the exception handling".

Regarding the "second" item (i.e., propagating or not), yes, I totally agree.

But regarding the "first" item (who should log it), conceptually, yes, but practically, no.

> It is totally fine for automation (or any other future caller) to call back to script to log it, because at least we then give automations the option not to.

First, the problem, even with the previous implementation, is that if any context switches happen in between (which, granted, doesn't happen now, but who knows in the future), then the "remembered exception" (inside the script helper) could get overwritten by another exception (from another run from another task.) And now that we're allowing multiple, completely independent runs, and that information is in the individual `_XxxScriptRun` object, but the caller only has access to the singular `Script` object, how will `Script.please_log_this_for_me` know _which_ run context (i.e., step # & action) to log the exception against?

Second, the script helper (specifically the `_XxxScriptRun` object) already logs several messages on "its own authority." Why should this be different? It knows how to properly log it, has the full context, and can do so without an otherwise unnecessary extra API.

I suppose one way to address both of these concerns is to decide based on whether or not the caller passes in a logger object. If so, log as before, including logging an exception. If not, don't log anything. But I'm not crazy with this idea because the other script helper log messages are useful.

So I suppose another way is to add a parameter to `Script.async_run` (that it passes down to the `_XxxScriptRun` object) that controls whether or not an exception should be logged. If so then do as it does now. If not then it becomes the caller's _full_ responsibility (i.e., there's no "hey, script helper, can you log this for me" method available, for the reasons I stated above.)

Personally, I still think the way it is (given addressing the propagation of the exception) is the best. The caller doesn't have to worry about whether or not, or how, to log the exception, because it will be done on its behalf. But if you still feel strongly otherwise, I'd go for adding a new parameter to `Script.async_run` and making it the caller's responsibility (and, I suppose, making the script helper _still_ log it if it doesn't propagate the exception.)

pnbruckner replied:
I looked at how to do this. It's not too bad. I'd add the new param to `Script.__init__` (with a default of `True`) because I think it'd be only one way for each type of caller. That would get passed down to `_XxxScriptRun.__init__` and ultimately to `_ScriptRunBase.__init__`. Then on each call to `_ScriptRunBase._async_step`, if an exception occurred, it would log it if 1) the user asked for it (via the new param), or 2) the exception is not getting propagated (which `_XxxScriptRun.async_run` would tell it.)

I'll go ahead and do this. If the user doesn't specify the new param it would work as it does now.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -121,6 +133,428 @@ class _SuspendScript(Exception):
+    async def _async_delay_step(self):
+        """Handle delay."""
+
+    def _start_delay_step(self):
```
This name is deceiving as we're not starting anything, we're returning the delay.

balloob replied:
Maybe prepare ?

pnbruckner replied:
Good point. I was thinking more "start of the delay/wait_template method" as opposed to "start of the delay/wait_template itself." That change will make it clearer. Will do.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -121,6 +133,428 @@ class _SuspendScript(Exception):
+    async def _async_wait_template_step(self):
+        """Handle a wait template."""
+
+    def _start_wait_template_step(self, async_script_wait):
```
same.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -121,6 +133,428 @@ class _SuspendScript(Exception):
+    async def async_stop(self) -> None:
+        """Stop script run."""
+        self._stop.set()
+        await self._stopped.wait()
```
This function should still cancel any ongoing delay/wait for template to be changed. Maybe we can have a `self._stoppable_task` variable that this method will cancel if it's set ?

pnbruckner replied:
It does stop them. Both `_async_delay_step` & `_async_wait_template_step` also monitor the `self._stop` Event.

balloob replied:
Nevermind this, just noticed your `async_delay_step` implementation!

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -121,6 +133,428 @@ class _SuspendScript(Exception):
+
+    async def _async_delay_step(self):
+        """Handle delay."""
+        await asyncio.wait(
```
Ooh I see what you did there. That's smart.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -121,6 +133,428 @@ class _SuspendScript(Exception):
+            done.set()
+
+        unsub = self._start_wait_template_step(async_script_wait)
+        if not unsub:
```
this can't happen ?

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -223,288 +666,59 @@ def referenced_entities(self):
+                self._raise("Already running")
+            if self._if_running == IF_RUNNING_RESTART:
+                self._log("Restarting script")
+                self.async_stop()
```
Should we await this?

balloob replied:
```suggestion
                await self.async_stop()
```

pnbruckner replied:
For sure. Missed that one. As you can probably tell, I haven't run many tests on the new code yet. I've done some ad hoc testing, and have written a few new tests, but I was waiting until the code settled down a bit before getting serious about testing the new code. 😅

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -223,288 +666,59 @@ def referenced_entities(self):
-        run_callback_threadsafe(self.hass.loop, self.async_stop).result()
+        self.last_triggered = utcnow()
+        if self._run_mode == RUN_MODE_LEGACY:
+            if self.is_running:
```
Feels a little weird that you're checking `self.is_running`, I know it's just a wrapped, but you could also do:
```suggestion
            if self._runs:
```

pnbruckner replied:
Yeah, that's probably a little clearer. Will do.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -121,6 +133,428 @@ class _SuspendScript(Exception):
+        if shared:
+            self._shared = shared
+        else:
+            # To implement legacy behavior we need to share the following "run state"
```
I thought that for shared we would just skip the rest of the delay? Did it actually still invoke after the delay was done?

pnbruckner replied:
Maybe the variable names or the comment need improvement to make what's happening here clearer.

Basically, the current (i.e., previously existing) implementation has _one_ `Script` object whose `async_run` method can be (and is, in the case of automations) called by separate (trigger) tasks.

The new "legacy" implementation still has just one `Script` object whose `async_run` method can be called multiple times, possibly "simultaneously." But to fit into the new class structure (required to truly support independent, simultaneous runs), there is one `_LegacyScriptRun` object _per call_ to `Script.async_run`, where they all share a common "run state" via the first `_LegacyScriptRun` object's `_current` and `_async_listeners` instance variables (accessed via the `_cur` & `_async_listener` properties.)

The ultimate goal is to make calls to `Script.async_run` -> `_LegacyScriptRun.async_run` behave as closely as possible to how calls to `Script.async_run` (by itself) used to behave.

And, to be clear, it's possible for the "user" (e.g., automation) to call `Script.async_run` (and hence, `_LegacyScriptRun.async_run`) when any previous "run" could be either still in the `async_run` method, or in a suspended state (from a `delay` or `wait_template`.)

I can tell you that it was a bit of a challenge to do this. But I'm feeling pretty good about the implementation now. It certainly passes the tests I created beforehand to verify the "legacy" behavior.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -223,288 +666,59 @@ def referenced_entities(self):
+                run = _BlockingScriptRun(self._hass, self, variables, context)
+        self._runs.append(run)
+        await run.async_run()
+        return True
```
why this ?

pnbruckner replied:
Sorry, not sure what you're referring to. Do you mean why the new return value?

If so, this is preparatory to the changes coming to the `automation` integration. Given the new `if_running` options, it will need to know if the call to `Script.async_run` actually started a new "run" so it can determine if it should update its `last_triggered` attribute. You'll notice that just above, if `self._if_running` is `IF_RUNNING_IGNORE` it returns `False`.

balloob replied:
Maybe we can put this in when we add it to the automation integration?

Not to get ahead on future PRs, but I think that a script deciding not to run should not impact if an automation triggered or not. It triggered, it's action just got ignored.

pnbruckner replied:
I can take it out for now.

But don't confuse the `script` integration and the script helper in this context. 😉 This return value (from the script helper) is telling the caller (i.e., `automation`) if the `action` "sequence" ran, not if that sequence happened to have a call to a `script` in it (that decided not to run.)

`homeassistant/helpers/script.py` should probably be called `homeassistant/helpers/sequence.py`, but that's probably too big of a change.

BTW, the automation's `last_triggered` attribute is already a misnomer, IMHO, and has caused a lot of confusion (as I've seen on the forum.) E.g., it is not updated if a `trigger` "fires" but a `condition` prevents the `action` sequence from running. It is only updated when the `action` sequence actually runs. It should probably be called `last_action_run`, but that would probably be too big of a change, too.

I was planning to keep the behavior of that attribute the same.

pnbruckner committed:
Change per review 2

- Don't propagate exceptions if call from user has already returned
  (i.e., for background runs or legacy runs that have suspended.)
- Allow user to specify if exceptions should be logged. They will still
  be logged regardless if exception is not propagated.
- Rename _start_script_delay and _start_wait_template_delay for
  clarity.
- Remove return value from Script.async_run.
- Fix missing await.
- Change call to self.is_running in Script.async_run to direct test of
  self._runs.

pnbruckner commented:
FYI I'm working on the tests now.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -130,39 +579,50 @@ def __init__(
+        if self._change_listener:
+            self._hass.async_add_job(self._change_listener)
+
+    def set_logger(self, logger):
```
Why was this added ?

pnbruckner replied:
While working on the previous PR, which included changes to script & automation, I wanted script and automation to be able to use their `object_id` in the creation of the logger:
```python
logging.getLogger(f"{__name__}.{split_entity_id(self.entity_id)[1]}")
```
This would result in something like:
```
homeassistant.components.automation.hello
```
as opposed to:
```
homeassistant.components.automation.automation.hello
```
But the `entity_id` is only known after `async_added_to_hass` is called, which is much later than when the `Script` object is created. So the API (of `Script`) needed to provide a way for its "user" to set the logger after `__init__` is called.

pnbruckner replied:
Actually, now that you mention it, and I think about it some more, it probably makes more sense to just remove `logger` from the `__init__` parameters, and make it a public member, initialized to `logging.getLogger(__name__)`. That way if the user doesn't care it doesn't have to do anything, but if it does, it can just do:
```python
script_obj = Script(...)
...
script_obj.logger = logging.getLogger(blah_blah)
```

balloob replied:
So let's remove it from this PR.

I rather have loggers be passed in as it forces the right logging namespace. In the automation PR we should delay making the script object until we have the entity ID.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -130,39 +579,50 @@ def __init__(
     def is_running(self) -> bool:
         """Return true if script is on."""
-        return self._cur != -1
+        return len(self._runs) > 0
```
Slightly faster.

```suggestion
        return bool(self._runs)
```

pnbruckner replied:
I had thought of doing that but decided using `len(x) > 0` was clearer. Also, I just tried both ways using `timeit` and, believe it or not, `len(x) > 0` is actually faster than `bool(x)`.

pnbruckner committed:
Change per review 3 and add tests

- Remove Script.set_logger().
- Enhance existing tests to check all run modes.
- Add tests for new features.
- Fix a few minor bugs found by tests.

balloob reviewed (approved):
Looks good to me! 🎉 

Anything else you want to change before we merge it?

pnbruckner commented:
Thanks! The reviews have been great and not only helped improve the code, but I learned a lot, too. 😃 

No, I think that's it. Besides the formal tests I also have a few scripts and automations I wrote specifically for this to test if the legacy behavior has been maintained, and they seem to indicate it has. So I think we're good.

Next PR will focus on using the new capabilities in the script integration. After that automations.
